### PR TITLE
Fix: show all series data on tooltip when multiple X-Axis

### PR DESCRIPTION
**Pull request by alexbalonperin**
## Description
Fixes https://github.com/recharts/recharts/issues/4757

I am using the state xAxisMap to access the domain of each x axis and create the missing payloads.
Not sure that's the best solution but it works and I couldn't really find a more elegant solution without major refactoring.

## How Has This Been Tested?
I have added a story for the line chart. Also, all automated tests are passing.

## Screenshots (if appropriate):
Before the fix:
Screenshot from 2024-07-12 12-58-46

After the fix:
Screenshot from 2024-07-12 13-00-02

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have added a storybook story or extended an existing story to show my changes

### CODE REVIEW

**Comment by ckifer** [comment]
Will need to make sure this doesn't break anything. Seems like it doesn't at first glance, but won't be good if it does - will also need to consider this in 3.x

**Comment by ckifer** [comment]
Also could you screenshot what this makes your TT look like and paste it in the PR? ty!

File: src/chart/generateCategoricalChart.tsx [hunk 1]
```diff
@@ -285,8 +285,19 @@ const getTooltipData = (state: CategoricalChartState, chartData: any[], layout:
   const activeIndex = calculateActiveTickIndex(pos, ticks, tooltipTicks, axis);
 
   if (activeIndex >= 0 && tooltipTicks) {
-    const activeLabel = tooltipTicks[activeIndex] && tooltipTicks[activeIndex].value;
-    const activePayload = getTooltipContent(state, chartData, activeIndex, activeLabel);
+    const activeLabel = tooltipTicks[activeIndex] && tooltipTicks[activeIndex].value; // for backward compatibility
+    let activePayload: any[];
+    if (state.xAxisMap) {
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:290]: thinking about this again, this does not work if you have a chart in the opposite layout.

Trying it, this causes a pretty big regression where the tooltip does not show at all

**Review by ckifer — CHANGES_REQUESTED** [review]
image
we can't do this unless we can satisfy all cases - or at least not break things

unit tests for both cases could help cover us

**Comment by ckifer** [comment]
code for the above breaking behavior
```jsx
      <ResponsiveContainer width="100%" height="100%">
        <LineChart width={500} height={300} layout="vertical">
          <CartesianGrid strokeDasharray="3 3" />

          <Tooltip />
          <Legend />
          {finDataOverTime.map((s, sIdx) => {
            return (
              <>
                <YAxis
                  key={`yaxis-${s.name}-`}
                  // domain={['auto', 'auto']}
                  orientation={yAxisOrientations[sIdx]}
                  dataKey={dataKey}
                  allowDuplicatedCategory={false}
                  yAxisId={sIdx}
                  type="category"
                />
                {categories.map((category, cIdx) => {
                  const dataRange = [].concat(
                    ...s.data.map(d =>
                      d[category]
                        ? [
                            {
                              [dataKey]: d[dataKey],
                              [category]: d[category],
                            },
                          ]
                        : [],
                    ),
                  );
                  return (
                    <>
                      <XAxis
                        key={s.name}
                        // dataKey={dataKey}
                        orientation={xAxisOrientations[cIdx]}
                        xAxisId={cIdx}
                        type="number"
                        domain={['auto', 'auto']}
                      />
                      <Line
                        key={`line-${s.name}-${category}`}
                        dataKey={category}
                        data={dataRange}
                        name={`${s.name}: ${category}`}
                        xAxisId={cIdx}
                        yAxisId={sIdx}
                        stroke={colors[cIdx]}
                        fill={colors[cIdx]}
                        strokeDasharray={`5 ${sIdx > 0 ? 5 : ''}`}
                      />
                    </>
                  );
                })}
              </>
            );
          })}
        </LineChart>
      </ResponsiveContainer>
```

File: src/chart/generateCategoricalChart.tsx (hunk 1 above)
Comment (alexbalonperin) [reply src/chart/generateCategoricalChart.tsx:290]: Nice catch. I didn't think of the vertical layout. Will add unit tests for both vertical and horizontal layouts.

Comment (alexbalonperin) [reply src/chart/generateCategoricalChart.tsx:290]: What's interesting is that the regular "Vertical" storybook seems to work fine and show the tooltip as expected. So it's only when there are multiple axis that things go wrong apparently.

I have been trying to write tests for the usecases that you mentioned but I am struggling (first time using react testing library). 

Here is what I was hoping to do:
- Copy the following test: "Renders tooltip when Tooltip item is added"
- Add extra x-axis, y-axis and area to mimic the storybook that you posted
- Fire a mouseover event on one of the area elements: `fireEvent.mouseOver(area)`
- Check for visibility of the tooltip wrapper: `expect(tooltipWrapper).toBeVisible()`

Unfortunately, this doesn't seem to work. I used the testing library screen.debug() to check the state of the dom but there is no change. Also, I added the storybook that you posted and I don't see any mouse event listener on any element so I am a bit confused. :thinking: 

I can't push broken tests so unless I copy the whole thing here you won't see my attempt.

Comment (ckifer) [reply src/chart/generateCategoricalChart.tsx:290]: You can push broken tests with --no-verify

File: test/chart/AreaChart.spec.tsx [hunk 2]
```diff
@@ -149,6 +150,182 @@ describe('AreaChart', () => {
...
+      const area = container.querySelectorAll('.recharts-surface')[0];
+
+      expect(tooltip).toHaveLength(1);
+      expect(tooltipWrapper).toHaveLength(1);
+
+      await waitFor(() => {
+        fireEvent.mouseOver(area);
+        expect(tooltipWrapper).toBeVisible();
+        screen.debug();
+      });
```
Comment (ckifer) [inline test/chart/AreaChart.spec.tsx:255]: For these you can consider re-using some utilities we have in place 

https://github.com/recharts/recharts/blob/979d921631364d33696d9a5b6f5fba561af5bf5e/test/component/Tooltip/tooltipTestHelpers.tsx#L17

https://github.com/recharts/recharts/blob/979d921631364d33696d9a5b6f5fba561af5bf5e/test/component/Tooltip/tooltipTestHelpers.tsx#L29

Comment (ckifer) [reply test/chart/AreaChart.spec.tsx:255]: and assert the payload with the second one

**Review by ckifer — COMMENTED** [review]
Just a few things, thanks for the patience

**Comment by ckifer** [comment]
Going to close this since this PR is to master, lets get this in 3.x instead

**Pull request closed without merging**